src/app/services/audio.service.ts

```diff
@@ -5,177 +5,281 @@ import * as moment from 'moment';
+          }
+          this.playNextTrack();
+        }
+      });
+  }
 
-    rewindTo(seconds: number) {
-        this.audioObj.currentTime = seconds;
-    }
+  private addEvents(obj, events, handler) {
+    events.forEach((event) => {
+      obj.addEventListener(event, handler);
+    });
+  }
 
-    playNextTrack() {
-        const currentTrack = this.trackList.find(track => track.id === this.audioID);
-        const nextTrack = this.trackList.find(track => track.track_number - 1 === currentTrack.track_number);
-        const isTrackListEnd = this.trackList.length === currentTrack.track_number + 1;
-        this.stop();
-        if (isTrackListEnd) return this.audioIDChange.next(null);
-        this.playStream(nextTrack);
-    }
+  private removeEvents(obj, events, handler) {
+    events.forEach((event) => {
+      obj.removeEventListener(event, handler);
+    });
+  }
 
-    formatTime(time: number, format: string = 'mm:ss') {
-        const momentTime = time * 1000;
-        return moment.utc(momentTime).format(format);
-    }
+  play() {
+    this.audioObj.play();
+  }
 
-    getState(): Observable<StreamState> {
-        return this.stateChange.asObservable();
-    }
+  pause() {
+    this.audioObj.pause();
+  }
 
-    getAudioID() {
-        return this.audioID;
-    }
+  stop() {
+    const volume = this.state.volume;
+    this.stop$.next();
+    this.state.volume = volume;
+  }
 
-    getAudioIDChange() {
-        return this.audioIDChange.asObservable();
-    }
+  rewindTo(seconds: number) {
+    this.audioObj.currentTime = seconds;
+  }
+
+  setVolume(volume: number) {
+    this.audioObj.volume = volume;
+    this.state.volume = volume;
+    this.stateChange.next(this.state);
+  }
 
-    getAlbumID() {
-        return this.albumId;
+  playNextTrack() {
+    if (this.isShuffled) {
+      return this.playStream(this.trackList[this.randomize()]);
+    }
+    const currentTrack = this.trackList.find(
+      (track) => track.id === this.audioID
+    );
+    const nextTrack = this.trackList.find(
+      (track) => track.track_number - 1 === currentTrack.track_number
+    );
+    const isTrackListEnd =
+      this.trackList.length === currentTrack.track_number + 1;
+    this.stop();
+    if (isTrackListEnd) {
+      this.audioIDChange.next(null);
+      this.state.playing = false;
+      this.state.paused = false;
+      return this.stateChange.next(this.state);
     }
+    this.playStream(nextTrack);
+  }
 
-    getTrackList() {
-        return this.trackList;
+  playPreviousTrack() {
+    const currentTrack = this.trackList.find(
+      (track) => track.id === this.audioID
+    );
+    const isTrackListStart = currentTrack.track_number - 1 < 0;
+    this.stop();
+    if (this.isShuffled) {
+      const currentTrackIndex = this.lastRandom.indexOf(
+        currentTrack.track_number
+      );
+      const prevRandomTrack = currentTrackIndex
+        ? this.trackList[this.lastRandom[currentTrackIndex - 1]]
```

Comment:
if `this.lastRandom` has an element `currentTrack.track_number` than `currentTrackIndex` will be assigned with index of that element. But if there is no such element - `-1` will be assigned to `currentTrackIndex`.
Than you check `currentTrackIndex` as a boolean value here. And this is a mistake because `-1` also gets `TRUE` and in your case `this.trackList[this.lastRandom[currentTrackIndex - 1]]` action always be chosen

Reply:
Thanks for pointing me on that, I've decided to use another shuffle algorithm